helper/querybuilder.php

```diff
@@ -0,0 +1,277 @@
+    /** @var array  */
+    protected $ns = [];
+    /** @var array */
+    protected $notns = [];
+    /** @var string */
+    protected $pid;
+    /** @var string */
+    protected $tagger = '';
+    /** @var string */
+    protected $limit = '';
+    /** @var string */
+    protected $where;
+    /** @var string */
+    protected $orderby;
+    /** @var string */
+    protected $groupby;
+    /** @var string */
+    protected $having = '';
+    /** @var array */
+    protected $values = [];
+
+    /**
+     * Shorthand method: calls the appropriate getter deduced from $this->field
+     *
+     * @return helper_plugin_tagging_querybuilder
+     */
+    public function getQuery()
+    {
+        if (!$this->field) {
+            throw new \RuntimeException('Failed to build a query, no field specified');
+        }
+        return ($this->field === 'pid') ? $this->getPages() : $this->getTags();
+    }
+
+    /**
+     * Processes all parts of the query for fetching tagged pages
+     *
+     * Returns the query builder object
+     *
+     * @return helper_plugin_tagging_querybuilder
+     */
+    public function getPages()
+    {
+        $this->where = $this->getWhere();
+        $this->groupby = 'pid';
+        $this->orderby = "cnt DESC, pid";
+        if ($this->tags && $this->logicalAnd) $this->having = ' HAVING cnt = ' . count($this->tags);
+
+        return $this;
+    }
+
+    /**
+     * Processes all parts of the query for fetching tags
+     *
+     * Returns the query builder object
+     *
+     * @return helper_plugin_tagging_querybuilder
+     */
+    public function getTags()
+    {
+        $this->where = $this->getWhere();
+        $this->groupby = 'CLEANTAG(tag)';
+        $this->orderby = 'CLEANTAG(tag)';
+
+        return $this;
+    }
+
+    /**
+     * Returns the SQL string
+     * @return string
+     */
+    public function getSql()
+    {
+        return $this->composeSql();
+    }
+
+    /**
+     * Returns the parameter values
+     * @return array
+     */
+    public function getParameterValues()
+    {
+        return $this->values;
+    }
+
+    /**
+     * Tags to search for
+     * @param array $tags
+     */
+    public function setTags($tags)
+    {
+        $this->tags = $tags;
+    }
+
+    /**
+     * Namespaces to limit search to
+     * @param array $ns
+     */
+    public function includeNS($ns)
+    {
+        $this->ns = $this->globNS($ns);
+    }
+
+    /**
+     * Namespaces to exclude from search
+     * @param array $ns
+     */
+    public function excludeNS($ns)
+    {
+        $this->notns = $this->globNS($ns);
+    }
+
+    /**
+     * Sets the logical operator used in tag search to AND
+     * @param bool $and
+     */
+    public function setLogicalAnd($and)
+    {
+        $this->logicalAnd = $and;
+    }
+
+    /**
+     * Result limit
+     * @param string $limit
+     */
+    public function setLimit($limit)
+    {
+        $this->limit = $limit ? " LIMIT $limit" : '';
```

Comment:
I would not build any SQL here, only set the internal variable and build SQL from it later. Also the limit should use a `?` placeholder for proper escaping.

Reply:
Done.